### DISCUSSION TITLE
Code Block

### DISCUSSION THREAD

**Discussion by markcaron** [RedHat-UX/red-hat-design-system#607] [General]
RH Code Block has been designed and is ready to be developed. But, there are a lot of features / API options.

[See Code Block mockups](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/)

Our ability to start with certain features might depend on how we implement `<rh-code-block>` (i.e. `zero-md`, Highlight.js, etc.), but I'd like to start this discussion about what features we should implement first, so that we don't bite off more than we can chew in our first version/release.

The MVP will include:

- [Base style for light theme](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/453bda61-1f4a-40e3-91f4-d7a8cfa5f656)
- [Base styles for dark theme](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/52b57bb1-5293-4d8f-8f7b-a5e7945fd220)
- Fixed width and fluid width

What additional features should we implement in the MVP?

1. [Compact size](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/9feae05c-4d8b-417a-b300-acf3a8a0cf17)
2. [Toolbar above code block](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/9feae05c-4d8b-417a-b300-acf3a8a0cf17)
3. [Copy to clipboard](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/d5a917c0-c3f4-424a-9747-dfb7e26431e9)
4. [Copy to clipboard in toolbar](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/6126c0c9-5986-4427-91d7-1e3d3831afba)
5. [Show language on hover](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/53a42cd9-7948-478e-b40e-a577f2a5a2aa)
6. [Show language in toolbar](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/26137bd9-6553-46cf-b246-85ef1d98c3c9)
7. [Syntax highlighting on light](https://xd.adobe.com/view/c4cc943e-618d-4fbe-8ab2-8e29fd135393-919f/screen/0cbc1f44-a1ee-4699-98a0-3b39bb1cfc8a) and [on dark](https://xd.adobe.com/view/c4cc943e-618d-4fbe-8ab2-8e29fd135393-919f/screen/77615b9c-d61c-42e7-babf-764cae0dcc1b) **
8. [Style toggle in toolbar](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/d5cf04de-c4c3-4908-988f-3c74e47c7a2c)
9. [Theme toggle in toolbar](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/ec3cb381-3cb4-4cd6-b4d7-c723bd812e35)
10. [Code language tabs](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/af4d9ac2-4645-434b-af6a-d1c96535b95c)
11. [Callout numbers](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/c587aadf-0b1c-4f6f-a3eb-bcd5c459c6d6) _(under review)_
12. [Line highlighting](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/34e36e58-265b-4ae9-9b04-b1e6f1646a7a)
13. [External links in toolbar](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/5118946d-f599-4363-9c19-84ea93346eda)
14. [Line wrapping and scrolling toggle](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/1c5c91e3-10fc-4fe9-a755-5bd720c969c0)
15. [Drag to resize](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/bd910221-dd0b-4ae6-9141-821aea34a857)
16. [Collapsible / expandable to full height](https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/screen/0b5a57e9-d786-4d7a-9b56-d94df6f24588)

** With syntax highlighting, we may also want to discuss what language(s) we should start with first.

Note: actual API names are not set in stone either, so feel free to help us decide how they should be named and implemented, of course.

**Comment by markcaron** [comment]
@RedHat-UX/developers @RedHat-UX/designers @RedHat-UX/a11y-unicorns ^

**Reply by bennypowers** [reply]
Thinking about that feature set, some of them are client-side only, like 3, 4, 8, 9, 14

but others like 2, 5, 6, 7, 11, 12 (especially syntax highlighting and themes) could be offered server side, which would be a better user experience. We would likely still need to offer them as opt-in client side features, e.g. in cases where the snippet is provided by third party scripts, unlexed.

**Comment by brianferry** [comment]
I feel like if I was designing around what I myself would prefer it would be the copy command whether it's an overlay or it's part of a toolbar.  The tabs and external links would be really nice for the docs but probably not MVP.  This all looks great btw loving these mocks
